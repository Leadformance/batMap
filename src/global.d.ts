// eslint-disable-next-line import/no-unresolved
import * as L from 'leaflet';

declare global {
  const L: L;
}
export {};
