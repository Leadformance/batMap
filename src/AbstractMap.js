'use strict';

require('./style.css');
const objectAssign = require('object-assign');
const dom = require('./utils/dom');

if (typeof Promise === 'undefined') {
    require('promise/lib/rejection-tracking').enable();
    window.Promise = require('promise/lib/es6-extensions.js');
}

class AbstractMap {
    constructor(domSelector, apiKey, locale, showCluster = false, showLabel = false, showPosition = false, callback = () => {}) {
        this.domElement = dom.isHTMLElement(domSelector) ? domSelector : document.querySelector(domSelector);
        this.domId = this.domElement.id || '';

        this.apiKey = apiKey;
        this.locale = locale || 'en';
        this.provider = '[No provider defined]';

        this.showCluster = showCluster;
        this.showLabel = showLabel;
        this.showPosition = showPosition;

        this.map = null;
        this.points = [];
        this.markers = [];
        this.icons = [];
        this.bounds = null;
        this.cluster = null;
        this.focusInProgress = false;

        this.defaultOptions = {
            zoom: 12,
            locationZoom: 16
        };

        this.load(callback);
    }

    load(callback) {
        callback();
    }

    setMapOptions(options = {}, markers = {}, labels = {}, clusters = {}) {
        this.mapOptions = objectAssign(this.defaultOptions, options);
        this.markersOptions = markers;
        this.labelsOptions = labels;
        this.clustersOptions = clusters;
    }

    getPoints() {
        return this.points;
    }

    getMarkers() {
        return this.markers;
    }

    getMarker(marker) {
        if (typeof marker === 'string') {
            marker = this.markers.find(m => {
                return m.id === marker;
            });
        }

        return marker;
    }

    getMarkerIcons() {
        return this.icons;
    }

    getMarkerIconByType(iconType) {
        if (this.icons[iconType]) {
            return this.icons[iconType];
        }

        return false;
    }

    getMarkerIconType(marker) {
        marker = this.getMarker(marker);
        return marker.iconType;
    }

    getGeolocation() {
        return new Promise((resolve, reject) => {
            if (navigator.geolocation) {
                navigator.geolocation.getCurrentPosition(resolve, reject, {
                    maximumAge: 60000,
                    timeout: 20000,
                    enableHighAccuracy: true
                });
            } else {
                reject(
                    new Error()
                );
            }
        });
    }

    initMap() {
        console.error(`${this.provider} has no 'initMap' method implemented.`);
    }

    setPoint() {
        console.error(`${this.provider} has no 'setPoint' method implemented.`);
    }

    clearPoints() {
        this.points.length = 0;
    }

    addMarkers() {
        console.error(`${this.provider} has no 'addMarkers' method implemented.`);
    }

    addUserMarker() {
        console.error(`${this.provider} has no 'addUserMarker' method implemented.`);
    }

    addMarker() {
        console.error(`${this.provider} has no 'addMarker' method implemented.`);
    }

    removeMarker() {
        console.error(`${this.provider} has no 'removeMarker' method implemented.`);
    }

    removeCluster() {
        console.error(`${this.provider} has no 'removeCluster' method implemented.`);
    }

    removeAllMarkers() {
        if (this.cluster) {
            this.removeCluster();
        }
        this.markers.forEach(marker => {
            this.removeMarker(marker);
        });
    }

    setMarkerIcons() {
        console.error(`${this.provider} has no 'setMarkerIcons' method implemented.`);
    }

    setIconOnMarker() {
        console.error(`${this.provider} has no 'setIconOnMarker' method implemented.`);
    }

    focusOnMarker() {
        console.error(`${this.provider} has no 'focusOnMarker' method implemented.`);
    }

    addCluster() {
        console.error(`${this.provider} has no 'addCluster' method implemented.`);
    }

    setCenter() {
        console.error(`${this.provider} has no 'setCenter' method implemented.`);
    }

    getCenterLatLng() {
        console.error(
            `${this.provider} has no 'getCenterLatLng' method implemented.`
        );
    }

    fitBounds() {
        console.error(`${this.provider} has no 'fitBounds' method implemented.`);
    }

    extendBounds() {
        console.error(`${this.provider} has no 'extendBounds' method implemented.`);
    }

    getBounds() {
        console.error(`${this.provider} has no 'getBounds' method implemented.`);
    }

    getBoundsLatLng() {
        console.error(`${this.provider} has no 'getBoundsLatLng' method implemented.`);
    }

    panTo() {
        console.error(`${this.provider} has no 'panTo' method implemented.`);
    }

    getZoom() {
        console.error(`${this.provider} has no 'getZoom' method implemented.`);
    }

    setZoom() {
        console.error(`${this.provider} has no 'setZoom' method implemented.`);
    }

    listenZoomChange() {
        console.error(`${this.provider} has no 'listenZoomChange' method implemented.`);
    }

    listenBoundsChange() {
        console.error(
            `${this.provider} has no 'listenBoundsChange' method implemented.`
        );
    }

    minifyMarkerIcons() {
        console.error(`${this.provider} has no 'minifyMarkerIcons' method implemented.`);
    }
}

module.exports = AbstractMap;
