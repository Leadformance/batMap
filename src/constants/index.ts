export * from './defaultConfig';
export * from './icons';
export * from './providers';
