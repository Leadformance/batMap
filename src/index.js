export { GoogleMap } from "./providers/gmaps/Gmaps";
export { GoogleMapWidget } from "./providers/gmaps-widget/GmapsWidget";
export { Leaflet } from "./providers/leaflet/Leaflet";
export { Mappy } from "./providers/mappy/Mappy";
export { MappyWidget } from "./providers/mappy-widget/MappyWidget";
