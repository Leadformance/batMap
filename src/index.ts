export { GoogleMaps } from './providers/gmaps';
export { GoogleMapsWidget } from './providers/gmaps-widget';
export { Leaflet } from './providers/leaflet';
export { Mappy } from './providers/mappy';
export { MappyWidget } from './providers/mappy-widget';
