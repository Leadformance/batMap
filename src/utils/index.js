export { domUtils } from './dom';
export { ieUtils } from './ie';
export { loaderUtils } from './loader';
export { isAString, isAFunction } from './type';
