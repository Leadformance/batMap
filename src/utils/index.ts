export { DomUtils } from './DomUtils';
export { LoaderUtils } from './LoaderUtils';
