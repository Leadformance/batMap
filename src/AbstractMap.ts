import './style.scss';
import { DEFAULT_CONFIG, Provider } from './constants';
import {
  ApiKey,
  Cluster,
  BatMapConfig,
  DefaultOptions,
  Icon,
  LocationPoint,
  MapObject,
  Marker,
} from './interfaces';
import { DomUtils } from './utils';

export class AbstractMap<T extends Provider> {
  protected readonly domElement: HTMLElement;
  protected readonly config: BatMapConfig<T>;

  protected points: LocationPoint<T>[] = [];
  protected markers: Marker<T>[] = [];
  protected icons: Icon<T>[] = [];
  protected focusInProgress = false;
  protected isMinifiedMarkerIcons = false;

  protected map?: MapObject<T>;
  protected cluster?: Cluster<T>;
  protected userMarker?: Marker<T>;

  constructor(
    domSelector: HTMLElement | string,
    protected readonly apiKey: ApiKey<T>,
    configuration: Partial<BatMapConfig<T>>,
    callback?: () => void,
    defaultOptions?: DefaultOptions<T>,
  ) {
    this.domElement = DomUtils.isHTMLElement(domSelector)
      ? (domSelector as HTMLElement)
      : document.querySelector(domSelector as string) || document.body;

    const config: BatMapConfig<T> = Object.assign(
      {},
      DEFAULT_CONFIG,
      configuration,
    );
    this.config = {
      ...config,
      mapOptions: Object.assign(
        {},
        defaultOptions?.mapOptions,
        config.mapOptions,
      ),
      markersOptions: Object.assign(
        {},
        defaultOptions?.markersOptions,
        config.markersOptions,
      ),
      labelsOptions: Object.assign(
        {},
        defaultOptions?.labelsOptions,
        config.labelsOptions,
      ),
    };

    this.load(callback);
  }

  load(callback?: () => void): void {
    callback && callback();
  }

  getPoints(): LocationPoint<T>[] {
    return this.points;
  }

  getMarkers(): Marker<T>[] {
    return this.markers;
  }

  getMarker(markerId: Marker<T> | string): Marker<T> | undefined {
    if (typeof markerId === 'string') {
      return this.markers.find(marker => marker.id === markerId);
    }

    return markerId;
  }

  getMarkerIcons(): Icon<T>[] {
    return this.icons;
  }

  getMarkerIconByType(iconType: string): Icon<T> | undefined {
    return this.icons.find(icon => icon.type === iconType);
  }

  getMarkerIconType(marker: Marker<T> | string): string | undefined {
    return this.getMarker(marker)?.iconType;
  }

  async getGeolocation(): Promise<GeolocationPosition> {
    if (!navigator.geolocation) {
      throw new Error('The browser does not support geolocation');
    }

    return new Promise((resolve, reject) => {
      navigator.geolocation.getCurrentPosition(resolve, reject, {
        maximumAge: 60000,
        timeout: 20000,
        enableHighAccuracy: true,
      });
    });
  }

  clearPoints(): void {
    this.points.length = 0;
  }

  removeAllMarkers(): void {
    this.removeCluster();

    this.markers.forEach(marker => {
      this.removeMarker(marker);
    });
  }

  removeCluster(): void {
    this.cluster && delete this.cluster;
  }

  removeMarker(markerId: Marker<T> | string): void {
    this.markers = this.markers.filter(
      marker => marker.id !== this.getMarker(markerId)?.id,
    );
  }
}
