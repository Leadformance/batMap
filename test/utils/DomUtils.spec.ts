import { fireEvent } from '@testing-library/dom';

import { DomUtils } from '../../src/utils';

describe('DomUtils', () => {
  const src = 'http://example.com/script.js';
  const href = 'http://example.com/style.css';

  afterEach(() => {
    document.head.innerHTML = '';
  });

  describe('addScript', () => {
    it('should add a script element into the document', () => {
      DomUtils.addScript(document.head, 'http://example.com/script.js');

      expect(document.head).toMatchSnapshot();
    });
  });

  describe('addStyle', () => {
    it('should add a link element into the document', () => {
      DomUtils.addStyle(document.head, 'http://example.com/style.css');

      expect(document.head).toMatchSnapshot();
    });
  });

  describe('addResources', () => {
    it('should only execute the callback function if the resources are empty', () => {
      const callback = jest.fn();
      DomUtils.addResources(document.head, [], callback);

      expect(callback).toHaveBeenCalledTimes(1);
      expect(document.head).toMatchSnapshot();
    });

    it('should add all resources into the document and execute the callback function once they are all loaded', () => {
      const style = document.createElement('link');
      style.href = href;

      const script = document.createElement('script');
      script.src = src;

      const callback = jest.fn();

      DomUtils.addResources(document.head, [style, script], callback);

      expect(document.head).toMatchSnapshot();
      expect(callback).toHaveBeenCalledTimes(0);

      fireEvent['load'](style);
      expect(callback).toHaveBeenCalledTimes(0);

      fireEvent['load'](script);
      expect(callback).toHaveBeenCalledTimes(1);
    });
  });

  describe('createScript', () => {
    it('should return a script element', () => {
      const script = DomUtils.createScript(src, false);

      expect(script).toMatchSnapshot();
      expect(script).toHaveProperty('async', false);
    });

    it('should return an async script element', () => {
      const script = DomUtils.createScript(src);

      expect(script).toMatchSnapshot();
      expect(script).toHaveProperty('async', true);
    });
  });

  describe('createStyle', () => {
    it('should return a link element with an href attribute', () => {
      expect(DomUtils.createStyle(href)).toMatchSnapshot();
    });
  });

  describe('isHTMLElement', () => {
    it('should return true with an HTML element', () => {
      expect(DomUtils.isHTMLElement(document.createElement('a'))).toEqual(true);
    });

    it('should return false with an object', () => {
      expect(DomUtils.isHTMLElement({ foo: 'bar' })).toEqual(false);
    });

    it('should return false with a string', () => {
      expect(DomUtils.isHTMLElement('foo')).toEqual(false);
    });
  });
});
