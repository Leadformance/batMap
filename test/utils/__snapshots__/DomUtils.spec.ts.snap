// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DomUtils addResources should add all resources into the document and execute the callback function once they are all loaded 1`] = `
<head>
  <link
    href="http://example.com/style.css"
  />
  <script
    src="http://example.com/script.js"
  />
</head>
`;

exports[`DomUtils addResources should only execute the callback function if the resources are empty 1`] = `<head />`;

exports[`DomUtils addScript should add a script element into the document 1`] = `
<head>
  <script
    src="http://example.com/script.js"
    type="text/javascript"
  />
</head>
`;

exports[`DomUtils addStyle should add a link element into the document 1`] = `
<head>
  <link
    href="http://example.com/style.css"
    rel="stylesheet"
  />
</head>
`;

exports[`DomUtils createScript should return a script element 1`] = `
<script
  src="http://example.com/script.js"
  type="text/javascript"
/>
`;

exports[`DomUtils createScript should return an async script element 1`] = `
<script
  src="http://example.com/script.js"
  type="text/javascript"
/>
`;

exports[`DomUtils createStyle should return a link element with an href attribute 1`] = `
<link
  href="http://example.com/style.css"
  rel="stylesheet"
/>
`;
